### ISSUE TITLE
Dependencies should consider annotations

### ISSUE THREAD

**Issue by codecholeric** [TNG/ArchUnit#136]
At the moment `JavaClass.getDirectDependenciesFromSelf()` and `JavaClass.getDirectDependenciesToSelf()` don't consider annotations.
Annotations (and their parameters) should be considered dependencies as well.
This includes

* all annotations on the `JavaClass` or any `JavaMember` of that class
* all `JavaClass` parameters of those annotations
* all `JavaClass` parameters of annotations within those annotations (and transitively their annotations and so on)

E.g.

```
@OnClass
@InAnnotation(
    type = AnotherDependency.class,
    nested = @InNestedAnnotation(type = YetAnotherDependency.class)
)
class SomeClass {
    @OnField
    Object field;

    @OnConstructor
    SomeClass() {}

    @OnMethod
    void method() {
    }
}
```

Then `OnClass`, `InAnnotation`, `AnotherDependency`, `InNestedAnnotation`, `YetAnotherDependency`, `OnField`, `OnConstructor` and `OnMethod` should all be returned as direct dependencies of `SomeClass`.

**Comment by slq** [comment]
Hi @codecholeric, have you started working on this issue? If not I can take it.

**Comment by codecholeric** [comment]
Hi @slq, first of all thanks a lot for the offer!! I've blocked this issue for someone who offered to contribute as well, so I assume it is in progress. Maybe you could look into #115 if you feel like it. I haven't marked it as "good start for contributing", but I think it would be the next best thing :wink: 
Would that be okay for you? I can of course support you, if any questions should arise!

**Comment by kosani** [comment]
The differences between JavaClass and JavaMemeber made me duplicate parts of the code in order to make proper Dependency descriptions. I think they might be solved by introducing / upgrading some new concepts.

Firstly, it seems to me that there might be a need for a HasDescription implementation on JavaClass.
It has already been implicitly used when creating a Dependency from origin and target:

```java
    String originType = origin.isInterface() ? "Interface" : "Class";
    String originDescription = originType + " " + bracketFormat(origin.getName());
```
might as well make it explicit, if there's not a better reason not to?

Secondly, the last part of the description has to do with the location that's derived differently:
JavaMemeber -> getOwner()
JavaClass -> this()

Should a JavaClass also implement HasOwner and return itself or is there a need for a new concept?

**Comment by codecholeric** [comment]
Yeah, `HasDescription` on `JavaClass` sounds reasonable, feel free to add that! But `HasOwner` I'm not convinced, I don't see any meaning in a class owning itself. I wouldn't add this just cause it's convenient at that one place...

**Comment by kosani** [comment]
I've managed to implement most **dependencies from** and that way seems to be pretty straight-forward. I've tried to fetch **dependencies to class** through `MemberDependenciesOnClass` (I might consider a different approach, since annotation is not a 'member'), but the problem is that the annotations are provided lazily through a supplier and the dependencies to the annotation do not get registered when `getDirectDependenciesToSelf` is invoked. I've tried using `Map<String, JavaAnnotation>` instead of `Supplier<Map<String, JavaAnnotation>>`, which works.. But I believe that the supplier was there for a reason, so I would like to know what's the reason is. 
Why are some `JavaClass` fields supplied (allFields, allMethods, allConstrucotrs), while others are used directly (fields, methods, constructors)?

**Comment by codecholeric** [comment]
The `Supplier` in general is used in places where the creation of certain objects

* has a noticeable performance impact
* is only used for specific tests

I don't know if this really holds for the simple building of annotations. You could compare an import of 50K classes though with and without the change of `Supplier` to direct reference and see, how a test that just tests package accesses compares in performance.

The `Supplier` is used for example, if a rule checks for dependencies to self, since then the graph must be completely resolved to find back references.

**Comment by kosani** [comment]
I've tried testing the performance with and without the supplier on a set of 27441 classes and the duration of both tests seem to converge to around 30s.

```java
ArchRule rule = classes().should().resideInAPackage("..");
JavaClasses classes = new ClassFileImporter().importPackages("java", "javax", "org", "com", "net", "junit", "sun", "jdk");
rule.check(classes);
```

So, if my testing and reasoning is correct, the supplier can be omitted in the case of annotations?

**Comment by codecholeric** [comment]
Yeah, I've checked it on the full JDK as well, and didn't notice any difference. Maybe originally I did something more complex there, but for now I think it's save to replace the `Supplier` with a simple map. The simpler the better :smiley:.
Would be cool, if you could throw it out of `JavaMember` then as well :wink:

**Comment by kosani** [comment]
While analyzing some failed tests I think I've figure out why annotations were supplied in a JavaClass and not instantiated eagerly in `completeMembers` method. This was just a mechanism to ensure that all `JavaClass.completeMembers` were called before any `JavaAnnotation` was created. For the same reason, some tests from `ClassFileImporterTest` failed, for example: `imports_class_with_annotation_with_empty_array`.

The test results in a `NullPointerException` when `annotation.get("primitives").get` is invoked because the `JavaAnnotation` of type `ClassAnnotationWithArrays` is built before `JavaClass.completeMembers` of `ClassAnnotationWithArrays` is invoked. Therefore, `primitives[]` was never built as a `JavaAnnotation` parameter.

I'm trying to find a third way of providing annotation dependencies to class that is not coupled with the `completeMembers` method. Is it possible to do it somewhere in `ClassGraphCreator.complete()` method? For instance, after `completeMembers()` is invoked?

Other tests fail mostly due to the new features added. For example, `Object()` is annotated with `jdk.internal.HotSpotIntrinsicCandidate`. I'll list these tests after these more critical ones are addressed.

**Comment by codecholeric** [comment]
Ay, now that you mention it, I do remember :wink: I think that might have been one of those places where a comment would have been a good idea.
I probably don't understand your problem completely though. Seems to me that `memberDependenciesOnClass` is created after `completeMembers` anyway, isn't it? Shouldn't you have all the info necessary by that time?

**Comment by kosani** [comment]
Only the info that is registered into the context. Since JavaAnnotations are instantiated via a `Supplier`, it's `context.createAnnotations(JavaClass.this)` doesn't get invoked by the time I start looking into `memberDependenciesOnClass`. Since registration into the context is done during `context.createAnnotations(JavaClass.this)`, the info about annotation dependencies to class is missing.

It looks like annotations need to be registered after `completeMembers` has been invoked on **all classes**, but also before `memberDependenciesOnClass` has been built **on any class**.

I've managed to do this in my latest commit, which fixed the issue, but there are still a couple of tests that are bugging me.

**Comment by codecholeric** [comment]
Ah I see! As far as I remember I wanted to avoid a second run over the classes there (since you could iterate all classes again and complete the annotations after the run to complete all members). So there are two things coming to my mind

a) check if that really costs so much performance (compare 50K classes or sth. like that again :wink:) If not you can iterate twice and still throw out the supplier
b) make the `memberDependenciesOnClass` reference a `Supplier` as well, so that the annotations are created when they are actually needed within the dependencies (similar to `accessesToSelf` as far as I remember, I think I avoided resolving the whole graph there eagerly as well)

As soon as I find some time I'll look more closely at your commits! Thanks a lot for your hard work!!

**Comment by kosani** [comment]
Okay, so it seems there is another issue with my current implementation of registering annotation dependencies to self. This time with annotations on fields and methods. I've tried to register annotations on members during registering fields, methods and constructors in `ClassGraphCreator`
```java
void registerFields(Set<JavaField> fields) {
    for (JavaField field : fields) {
        fieldTypeDependencies.put(field.getRawType(), field);
        registerMemberWithAnnotations(field);
    }
}
```
which is also faulty because I am accessing the supplier of annotations (eagerly) before all methods have been registered. Thus, the following tests are broken 
```
ClassFileImporterTest.imports_fields_with_complex_annotations_correctly
ClassFileImporterTest.imports_fields_with_two_annotations_correctly
ClassFileImporterTest.imports_method_with_annotation_with_empty_array
```
for the same reason `imports_class_with_annotation_with_empty_array` was broken.

It seems that I also need to access all JavaMembers with annotations, but only after each classes members have been completed.

**Comment by kosani** [comment]
I've tackled the issue above by invoking a new `completeAnnotations` method after `compelteMembers` in `ClassGraphCreator`. I ran the same speed test from earlier and didn't see any difference.

Most of the tests pass, except a few that are mainly due to the new feature. Namely,
```
ShouldClassesThatTest.areNotMetaAnnotatedWith_typeName[1: com.tngtech.archunit.lang.syntax.ClassesThatInternal@72e5a8e]
ShouldClassesThatTest.only_areMetaAnnotatedWith_predicate[1: com.tngtech.archunit.lang.syntax.ClassesThatInternal@73e9cf30]
ShouldClassesThatTest.only_areMetaAnnotatedWith_type[1: com.tngtech.archunit.lang.syntax.ClassesThatInternal@942a29c]
ShouldClassesThatTest.areNotMetaAnnotatedWith_predicate[1: com.tngtech.archunit.lang.syntax.ClassesThatInternal@3012646b]
ShouldClassesThatTest.only_areMetaAnnotatedWith_typeName[1: com.tngtech.archunit.lang.syntax.ClassesThatInternal@117159c0]
ShouldClassesThatTest.areNotMetaAnnotatedWith_type[1: com.tngtech.archunit.lang.syntax.ClassesThatInternal@68c9133c]
ShouldOnlyByClassesThatTest.areNotMetaAnnotatedWith_typeName[1: com.tngtech.archunit.lang.syntax.ClassesThatInternal@6440112d]
ShouldOnlyByClassesThatTest.areNotMetaAnnotatedWith_predicate[1: com.tngtech.archunit.lang.syntax.ClassesThatInternal@4d1c005e]
ShouldOnlyByClassesThatTest.areNotMetaAnnotatedWith_type[1: com.tngtech.archunit.lang.syntax.ClassesThatInternal@6f204a1a]
```
due to the fact that a new failure is now relevant: 
```
Class <com.tngtech.archunit.lang.syntax.elements.ShouldOnlyByClassesThatTest$MetaAnnotatedClass> is annotated with <com.tngtech.archunit.lang.syntax.elements.ShouldOnlyByClassesThatTest$MetaAnnotatedAnnotation> in (ShouldOnlyByClassesThatTest.java:0)
```
so the `ClassesShouldEvaluator.filterClassesAppearingInFailureReport` method also filters the `MetaAnnotatedAnnotatio` which isn't present in the assertions.

I am not sure what to do about these tests. Adding the class to the assertion will make the  `classes().should().onlyBeAccessed().byClassesThat()` part then fail.

**Comment by codecholeric** [comment]
I think you have to split those tests into two tests. The tests mostly assume that everything will behave the same for `accessClassesThat()` and `dependOnClassesThat()` which is not true anymore after your adjustments.
So all failing tests IMHO need to be split so you test once with `accessClassesThat()` which should stay the old way and once with `dependOnClassesThat()` where correctly the new violation by `MetaAnnotatedAnnotation` should appear.
You can mix dataprovider tests (the ones with `@UseDataProvider`) and normal tests freely, so it should be no problem to just remove the `@UseDataProvider` and the parameter, duplicate the test and directly check it on the respective rule start from the `@DataProvider` method. (I don't think you can DRY too much here, cause it's only 5 lines and input as well as expected output are different)

**Comment by t-h-e** [comment]
I know everyone hates this question, but may I ask what the current status is?
I would also be willing to help (although I have not worked on ArchUnit before), as I would really like to have this feature as well :)

**Comment by kosani** [comment]
Hi @t-h-e 
I've submitted a pull request #171 and am waiting for approval. @codecholeric any new info on this pull request?

**Comment by codecholeric** [comment]
I know, this is totally on me :frowning_face: 
I had problems over the recent months to find enough time, but now I'll push through getting these PRs merged and create a new release. Just give me a couple of days guys :wink:

**Comment by jbduncan** [comment]
@codecholeric I'm a bit lost, because I've just seen that this issue has been closed, but I'm struggling to find any commits in `master` that point back to this issue. So I wondered if you could clarify the state of things for me?

**Comment by codecholeric** [comment]
@jbduncan Oh, sorry, I guess the "resolves" part had been written in code and thus was not linked. Yes, #171 should solve this, i.e. annotation types, classes referenced from annotations and enum parameter types should not be detected as annotations in both directions :smiley:

**Comment by jbduncan** [comment]
@codecholeric Oh brilliant! I'm glad to see that #171 was merged in recently. Very well done to both you and @kosani for resolving this, it looks like it took a _lot_ of time and effort. 👍

**Comment by codecholeric** [comment]
Yes, it was a lot harder than expected, I made a big misjudgement there. In particular because of fun things like

```
@Funny(
    how = [
        @How(deeply = Deeply.INSTANCE),
        @How(veryDeeply = @EvenDeeper(type = YouCanNestAnnotations.class))
    ]
)
```

**Comment by uhafner** [comment]
I think one case has been forgotten: Annotations for method parameters are not recognized by the fix. 

Example: I want to forbid that the annotation `@CheckForNull` in my project. In the following method only the method annotation will be detected, but not the parameter annotation.
```
    @CheckForNull
    public boolean exists(@CheckForNull final String fileName) {
```

I'm using the following check:
```
    @ArchTest
    static final ArchRule NO_FORBIDDEN_ANNOTATION_CALLED 
          = noClasses().should().dependOnClassesThat()
                .haveFullyQualifiedName(CheckForNull.class.getName());
```

Or is this possible using a different approach?

**Comment by nbrugger-tgm** [comment]
For me this still doesn't work:
```java
@Override
		@RequiresPermission(Permission.VIEW_COLUMN)
		public List<FormulaDto> getFormulas() {
			....
		}
```
with this rules:
```java
classes().should(not(new ArchCondition<JavaClass>("access @NoUse stuff") {
			@Override
			public void check(JavaClass item, ConditionEvents events) {
				... item.getAllAccessesFromSelf() ...
			}
		}));
```
and `item.getAccessesFromSelf()` does not return any access to `RequiresPermission` or `Permission`.
`RequiresPermission` is a RUNTIME scoped annotation. 

The other way around also does not work:
```java
noMembers().that().areAnnotatedWith(Permission.NoUse.class).should(new ArchCondition<JavaMember>("be accessed") {
			@Override
			public void check(JavaMember item, ConditionEvents events) {
				... item.getAccessesToSelf() ...
			}
		});
```
when item is `Permission.VIEW_COLUMN` , `item.getAccessesToSelf()` does not return anything around `getFormulas()`
